caching: Support out-of-thread (async) HttpCache implementations

yosrym93 opened the pull request:
Commit Message:
Added support to out-of-thread (async) implementations of HttpCache.
Signed-off-by: Yosry Ahmed <yosryahmed@google.com>

Additional Description:
- Cache callbacks are now posted to the dispatcher to make sure they are run on the worker thread.
- Simplified the CacheFilter state management and tests as all caches are now treated in the same way.

Risk Level: Low
Testing: Updated existing tests
Docs Changes: N/A
Release Notes: N/A

toddmgreer commented on source/extensions/filters/http/cache/cache_filter.cc:
```diff
@@ -113,11 +102,12 @@ Http::FilterHeadersStatus CacheFilter::encodeHeaders(Http::ResponseHeaderMap& he
   }
-  if (filter_state_ == FilterState::WaitingForCacheBody) {
-    // Encoding stream stopped waiting for cached body (and trailers) to be encoded.
+  if (filter_state_ == FilterState::EncodeServingFromCache) {
```
Why wouldn't we return Continue here? It will let they bytes flow through downstream filters and to the client, right?

yosrym93 replied:
My understanding is that if we do not stop iteration, then the downstream filters will start processing the response immediately after `encodeHeaders` returns. The body may have not been fetched entirely from the cache yet. We need to fetch the complete cached body before we let the downstream filters process the response.

Actually, probably the `onBody` will be posted to the dispatcher as the last event, after all the filters finish `encodeHeaders` and `encodeData`. I think we need to stop the response encoding stream, add the cached response, then continue the encoding stream.

yosrym93 replied:
I did not notice that this is `encodeData` not `encodeHeaders`, sorry.

When `encodeHeaders` returns `StopIteration`, this does not stop `encodeData` of the same filter from being called. So if the 304 response includes a body by mistake and `encodeData` is called, we need to make sure that we keep the encoding stream stopped if the cached body is not fully fetched yet.

For some reason, we cannot return `StopAllIterationAndWatermark` from `encodeHeaders`. The documentation suggests this as it states that `continueDecoding` rather than `continueDecoding/continueEncoding` must be called to continue. I also tried returning `StopAlliterationAndWatermark` and the integration test crashed.

Maybe we should actually return `StopIterationNoBuffer` here to drop any existing 304 body? I am not sure this is what it does though.

@mattklein123 can you please confirm or correct the info in this comment?

toddmgreer replied:
We want downstream filters to start processing the response immediately. As soon as we have enough data to fill a packet, we want to send it on to the client, both so the client can start processing it, and so we can deallocate buffered bytes more quickly to keep our RAM usage under control.

yosrym93 replied:
@toddmgreer I updated my comment, I thought the comment was about `encodeHeaders` not `encodeData`, sorry.

yosrym93 replied:
I understand what you mean, but I am not sure that we can do this with `addEncodedData`. Especially that there is no way to tell it when the stream ends. What I understood from looking at the code is that we should stop iteration, add the cached data, then continue iteration.

Maybe we should use `injectDecodedDataToFilterChain` instead? But I am not exactly sure how we are supposed to use it.

yosrym93 replied:
Also, `addEncodedData` does not cause `encodeData` of the same filter to be called anyway.

yosrym93 replied:
It looks like to do this we will need to switch to `injectEncodedDataToFilterChain`, which is only used by the fault filter atm. Maybe postpone this to a separate PR? It wasn't introduced by this PR anyway.

toddmgreer commented on test/extensions/filters/http/cache/cache_filter_test.cc:
```diff
@@ -28,8 +29,87 @@ class CacheFilterTest : public ::testing::Test {
 
+  void SetUp() override {
+    api_ = Api::createApiForTest();
+    dispatcher_ = api_->allocateDispatcher("test_thread");
```
api_ and dispatcher_ could be initialized in their declarations.

yosrym93 commented on source/extensions/filters/http/cache/cache_filter.cc:
```diff
@@ -129,23 +119,61 @@ Http::FilterDataStatus CacheFilter::encodeData(Buffer::Instance& data, bool end_
 
+void CacheFilter::getHeaders(Http::RequestHeaderMap& request_headers) {
+  ASSERT(lookup_, "CacheFilter is trying to call getHeaders with no LookupContext");
+  lookup_->getHeaders([this, &request_headers,
```
I didn't know about `StreamInfo`, thanks!

yosrym93 replied:
Actually `getRequestHeaders()` returns a pointer to `const RequestHeadersMap`, so we can't use it here.

toddmgreer replied:
Hadn't noticed that. Please resolve this comment.

mattklein123 commented on source/extensions/filters/http/cache/cache_filter.cc:
```diff
@@ -32,8 +33,12 @@ CacheFilter::CacheFilter(const envoy::extensions::filters::http::cache::v3alpha:
 void CacheFilter::onDestroy() {
-  lookup_ = nullptr;
-  insert_ = nullptr;
+  if (lookup_) {
```
I haven't tracked this PR, but no, this is not guaranteed. Can you describe a bit more what is going on or do you want me to review the entire PR?

toddmgreer replied:
Most of CacheFilter's plugin cache implementations will complete requests asynchronously, and there's no guarantee of what thread those completions will run on, so their callbacks get posted to our dispatcher. Suppose a plugin posts a callback, then (perhaps due to a client disconnect) CacheFilter::onDestroy is called. That callback must happen before the CacheFilter gets deleted (or else it will access a deleted CacheFilter). AFAICT, the current code does give us the correct order (because it runs pending callbacks before it runs deferred deletion), and I'll be surprised if it ever needs to change, but we need to make sure.

I think that all we need to do is add documentation and tests to lock in the current behavior of running pending callbacks before deferred deletion.

If we can't rely on this, we'll have to go back to a prior iteration of the original CacheFilter PR, and use either weak_ptrs or shared_ptrs to the CacheFilter, but that complicates the code and adds bus-locked operations.

mattklein123 replied:
The solution here is to not use post, but use a "timer" which can be cancelled. Can we switch to that?

mattklein123 replied:
Sorry nevermind, I didn't read the cross-thread thing. So, yes, right now the problem with post is it can't be cancelled. This has come up multiple times, and is also related to https://github.com/envoyproxy/envoy/issues/12364#issuecomment-674582595 which I'm about to fix because it has bitten us multiple times. I think other than the weak_ptr/shared_ptr dance, the other option is to build this into post itself and make post cancellable. This would require infra work before your change though.

mattklein123 replied:
> f we can't rely on this, we'll have to go back to a prior iteration of the original CacheFilter PR, and use either weak_ptrs or shared_ptrs to the CacheFilter, but that complicates the code and adds bus-locked operations.

I don't believe this can be relied on in the current implementation.

toddmgreer replied:
After taking another look at the current implementation, I think I see what you were alluding to. If someone calls deferredDelete (perhaps on something unrelated to us), then the call to post happens, then CacheFilter::onDestroy; then the deletion could happen before the post gets run, and we're in trouble.

We could easily guarantee that callbacks posted before deferredDelete(foo) is called are run before foo is deleted.  The simplest way I see is to have deferred_delete_cb_'s lambda call runPostCallbacks() before it calls clearDeferredDeleteList(). Would such a change would be useful outside of CacheFilter?

jmarantz replied:
Suggestion: rather than typing the asynchronous completion of a caching operation to a filter's lifetime, can we tie it to the config or something else? It seems that holding onto all the state of a filter is not necessary to complete (say) a cache insert.

Of course an http cache lookup should be tied to a filter as that's only in service of a request.

mattklein123 replied:
> We could easily guarantee that callbacks posted before deferredDelete(foo) is called are run before foo is deleted. The simplest way I see is to have deferred_delete_cb_'s lambda call runPostCallbacks() before it calls clearDeferredDeleteList(). Would such a change would be useful outside of CacheFilter?

It's something we can think about, though I'm hesitant to build more assumptions like this into the dispatcher code as it becomes fragile to maintain. We would definitely want to discuss that in a separate PR. I would honestly be more inclined to just implement a cancellable post. I think it's a cleaner interface, would fix your problem, and has come up multiple times before.

> Suggestion: rather than typing the asynchronous completion of a caching operation to a filter's lifetime, can we tie it to the config or something else? It seems that holding onto all the state of a filter is not necessary to complete (say) a cache insert.

Yeah this is another possible design where you have a cache manager that the filter defers to for this sort of thing.

yosrym93 replied:
@mattklein123 I am following the discussion at #12364. Can you elaborate how is it related to our problem here?

jmarantz replied:
More thoughts: if the client closes the connection during an outstanding lookup, we would need to cancel the cache lookup, and note in its state that the request-related data structures referenced in the lookup context are defunct and can no longer be accessed. That way you can allow the request to be retired and destroyed, even if there is an outstanding async cache lookup.

I think this is pretty do-able, having done similar flows in the past for other projects.

yosrym93 replied:
@jmarantz the problem arises if the cache lookup has already finished and posted the callback to the dispatcher, but then the client closes the connection and the filter chain is destroyed. Currently, there's no guarantee that the posted callback will run before the filter is deleted. If the filter is deleted first, the posted callback will run on a destroyed filter.
One solution here is to capture a weak_ptr to the CacheFilter in the posted callback, and check that the CacheFilter is still alive before accessing any of its members/methods.

jmarantz replied:
Can we arrange a destroy-notify callback from the filter chain? The cache system could then mark the filter structure as stale and avoid referencing it further.

jmarantz replied:
of course weak_ptr does the job too, if the thing being pointed to is already a shared_ptr.

jmarantz replied:
one more thing: the advantage of a callback over a weak_ptr is that you might be able to actively cancel an outstanding lookup request (if that is possible).

mattklein123 replied:
> @mattklein123 I am following the discussion at #12364. Can you elaborate how is it related to our problem here?

It's loosely related in the sense that we have TLS post/callbacks that are being running beyond the acceptable lifetime.

> one more thing: the advantage of a callback over a weak_ptr is that you might be able to actively cancel an outstanding lookup request (if that is possible).

This is how a cancellable post would be accomplished within dispatcher. This is what I recommend if you want a core solution, otherwise you can implement this yourself.

mattklein123 commented:
I assigned myself so can help review the entire PR once we sort out the implementation plan.

/wait

toddmgreer commented:
Calling onDestroy on lookup_ and insert_ takes care of the cancellation
callback. Holding a weak_ptr<CacheFilter> in the callback lambda takes care
of CacheFilter getting deleted before the post happens. (See PR #7198 for
an earlier state of this approach.)

On Tue, Aug 18, 2020 at 1:49 PM Matt Klein <notifications@github.com> wrote:

> *@mattklein123* commented on this pull request.
> ------------------------------
>
> In source/extensions/filters/http/cache/cache_filter.cc
> <https://github.com/envoyproxy/envoy/pull/12622#discussion_r472484140>:
>
> > @@ -32,8 +33,12 @@ CacheFilter::CacheFilter(const envoy::extensions::filters::http::cache::v3alpha:
>      : time_source_(time_source), cache_(http_cache) {}
>
>  void CacheFilter::onDestroy() {
> -  lookup_ = nullptr;
> -  insert_ = nullptr;
> +  if (lookup_) {
>
> @mattklein123 <https://github.com/mattklein123> I am following the
> discussion at #12364 <https://github.com/envoyproxy/envoy/issues/12364>.
> Can you elaborate how is it related to our problem here?
>
> It's loosely related in the sense that we have TLS post/callbacks that are
> being running beyond the acceptable lifetime.
>
> one more thing: the advantage of a callback over a weak_ptr is that you
> might be able to actively cancel an outstanding lookup request (if that is
> possible).
>
> This is how a cancellable post would be accomplished within dispatcher.
> This is what I recommend if you want a core solution, otherwise you can
> implement this yourself.
>
> —
> You are receiving this because you were assigned.
> Reply to this email directly, view it on GitHub
> <https://github.com/envoyproxy/envoy/pull/12622#discussion_r472484140>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AFRAWPPRY7CMV7FFNGKPGF3SBLSLXANCNFSM4P5NVFGQ>
> .
>

yosrym93 commented:
I added safeguard against the "callback posted then filter deleted" case we have been discussing by using a weak_ptr. I tried to make it as simple as possible. I added a TODO to look into other solutions as they arise (guaranteed dispatcher ordering of posts and deletions, cancellable posts, etc.).
I also added a test for this that fails if the filter is accessed after being deleted when run with the ASAN sanitizer (otherwise the behavior is naturally undefined - it may give false positives).
Looking forward for everyone to take a look!

mattklein123 commented on source/extensions/filters/http/cache/cache_filter.cc:
```diff
@@ -129,23 +120,103 @@ Http::FilterDataStatus CacheFilter::encodeData(Buffer::Instance& data, bool end_
+            cache_filter->onHeaders(LookupResult{status, absl::WrapUnique(headers), content_length,
+                                                 response_ranges, has_trailers},
+                                    request_headers);
+          }
```
If this is false you are going to leak headers I think. You need to wrap it before the lock. Please add a test for this. You may want to consider thread_synchronizer if you want real threads, or you can do it with mocks.

yosrym93 replied:
Good catch! Thanks for pointing this out.
The unit test I added in the last commit (in `cache_filter_test`) should be already exercising the case where the callback is made after the filter is deleted, which should be the case where the headers leak. Is there some kind of leak detector tests that run in the CI? What's Envoy's policy on testing for memory leaks?

mattklein123 replied:
The heap checker runs by default and should check for leaks. I'm not sure why it's not catching this case but it should be.

yosrym93 replied:
Is there a way to run it locally so that I can debug this?

yosrym93 replied:
I wonder why it didn't catch this, I will take another look at the test.

yosrym93 replied:
I fixed the leak but I don't know how the heap checker was not catching a memory leak before the fix. Maybe something is wrong with my test but I can't see it.

mattklein123 replied:
Can you verify the leak checker works by leaking something else obvious, and then checking your test? I'm also concerned that your test is not testing what you think it is.

yosrym93 replied:
Interesting observation, when I add a dummy leak like this:
```
dispatcher.post([self, &request_headers, status = result.cache_entry_status_,
                     headers_raw_ptr = result.headers_.release(),
                     response_ranges = std::move(result.response_ranges_),
                     content_length = result.content_length_,
                     has_trailers = result.has_trailers_,
                     dummy = new int]() mutable {
      // Wrap the raw pointer in a unique_ptr before checking to avoid memory leaks.
      if (CacheFilterSharedPtr cache_filter = self.lock()) {
        cache_filter->onHeaders(
            LookupResult{status, absl::WrapUnique(headers_raw_ptr), content_length, response_ranges, has_trailers},
            request_headers);
        absl::WrapUnique(dummy);
      }
    });
```

My test cause the heap checker to detect the leak of `dummy`, but not `headers_raw_ptr`.

mattklein123 commented on source/extensions/filters/http/cache/cache_filter.cc:
```diff
@@ -129,23 +120,103 @@ Http::FilterDataStatus CacheFilter::encodeData(Buffer::Instance& data, bool end_
+    // wrapped in a unique_ptr inside the lambda.
+    dispatcher.post([self, body = body.release()] {
+      if (auto cache_filter = self.lock()) {
+        cache_filter->onBody(absl::WrapUnique(body));
```
Same, you will leak

mattklein123 commented on source/extensions/filters/http/cache/cache_filter.cc:
```diff
@@ -129,23 +120,103 @@ Http::FilterDataStatus CacheFilter::encodeData(Buffer::Instance& data, bool end_
+    // pointer then wrapped in a unique_ptr inside the lambda.
+    dispatcher.post([self, trailers = trailers.release()] {
+      if (auto cache_filter = self.lock()) {
+        cache_filter->onTrailers(absl::WrapUnique(trailers));
```
Leak

mattklein123 commented on source/extensions/filters/http/cache/http_cache.h:
```diff
@@ -241,6 +241,14 @@ class InsertContext {
   // Inserts trailers into the cache.
   virtual void insertTrailers(const Http::ResponseTrailerMap& trailers) PURE;
 
+  // This routine is called prior to an InsertContext being destroyed. InsertContext is responsible
+  // for making sure that any async events are cleaned up in the context of this routine. This
+  // includes timers, network calls, etc. The reason there is an onDestroy() method vs. doing this
+  // type of cleanup in the destructor is due to the deferred deletion model that Envoy uses to
+  // avoid stack unwind complications. InsertContext must not invoke any callbacks after having
+  // onDestroy() invoked.
+  virtual void onDestroy() PURE;
```
I'm not completely sure why we need onDestroy() here and below. This should never be called in the context of a cache callback, so why can't it be handled by the relevant destructors?

toddmgreer replied:
My design originally didn't include (Insert|Lookup)Context::onDestroy, but I discovered that it was necessary in order to avoid potential data races. Without onDestroy, consider the following sequence with a hypothetical plugin-provided RpcInsertContext that implements InsertContext:

1. [Filter's thread] CacheFilter calls InsertContext::insertBody.
2. [Filter's thread] RpcInsertContext sends RPC and returns.
3. [Filter's thread] Client disconnects; CacheFilter dtor begins
4. [Filter's thread] RpcInsertContext dtor begins
5. [Other thread] RPC completes and calls RpcInsertContext::onRpcDone.
RpcInsertContext's dtor and onRpcDone cause a data race in RpcInsertContext.

There are several ways we could prevent this:
1. Document that all access to classes that inherit from InsertContext must be on the Filter's thread. This complicates the requirements on plugins. Mistakes will be painful. I recommend against this approach.
2. Share ownership of InsertContext. Some plugins would give shared_ptrs to their callbacks (to keep InsertContext alive). Plugins that want to cancel aync operations would only hold weak_ptrs. When CacheFilter gets deleted and drops its InsertContextSharedPtr, RpcInsertContext's dtor would cancel any outstanding async operations. Anything that completes after that would bail when it fails to convert the weak_ptr to a shared_ptr.
3. Call InsertContext::onDestroy in CacheFilter::onDestroy. RpcInsertContext::onDestroy cancels any outstanding async operations. If necessary, it blocks on that cancellation.

I chose onDestroy because an earlier shared ownership approach received unfavorable reviews, and because onDestroy is an established Envoy idiom. I consider both shared ownership and onDestroy to be reasonable approaches. onDestroy leaves more room for implementation mistakes, but it's more explicit, and avoids shared ownership.

mattklein123 replied:
> I chose onDestroy because an earlier shared ownership approach received unfavorable reviews, and because onDestroy is an established Envoy idiom. I consider both shared ownership and onDestroy to be reasonable approaches. onDestroy leaves more room for implementation mistakes, but it's more explicit, and avoids shared ownership.

Right, OK. I vaguely recall us having a conversation on this last time. Sorry for making you type this out again. Would it be worth it to take this nice detailed comment and put it in the code somewhere so we have this history and I don't end up asking this again in a few months? I think it's worthwhile, even if slightly slimmed down. Up to you.

yosrym93 replied:
@toddmgreer Happy to make any additions/modifications to these functions documentation as you see fit.

yosrym93 commented:
@mattklein123 I fixed the test. Turns out the lookup in the test did not find any cache entries so `headers` was an empty `unique_ptr`, that's why there were no leaks. I fixed the test and now it fails when the pointer is wrapped inside the `if` condition.

mattklein123 commented:
> @mattklein123 I fixed the test. Turns out the lookup in the test did not find any cache entries so headers was an empty unique_ptr, that's why there were no leaks. I fixed the test and now it fails when the pointer is wrapped inside the if condition.

Great!
